src/environment.c

```diff
@@ -0,0 +1,137 @@
+#include <sys/queue.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <stdio.h>
+#include "environment.h"
+
+static SLIST_HEAD( csh_env_llist, csh_env_entry ) csh_env = SLIST_HEAD_INITIALIZER(csh_env);
+struct csh_env_entry {
+    char *name;
+    char *value;
+    SLIST_ENTRY(csh_env_entry) list;
+};
+
+static struct csh_env_entry *csh_get_env_entry(const char *name) {
+    struct csh_env_entry *cur;
+    SLIST_FOREACH(cur, &csh_env, list) {
+        if(strcmp(name, cur->name) == 0) {
+            return cur;
+        }
+    }
+    return NULL;
+}
+
+ __attribute__ ((malloc)) static char *csh_copy_char_and_grow(char *dest, size_t at, size_t *dest_len, char c) {
+    if (at == *dest_len) {
+        *dest_len += 16;
+        dest = realloc(dest, *dest_len);
+    }
+    dest[at] = c;
+    return dest;
+}
+
+void csh_foreach_var(csh_foreach_var_cb cb) {
+    struct csh_env_entry *cur;
+    SLIST_FOREACH(cur, &csh_env, list) {
+        cb(cur->name);
+    }
+}
+
+char *csh_getvar(const char *name) {
+    char *result = NULL;
+    struct csh_env_entry *cur = csh_get_env_entry(name);
+    if (NULL != cur) {
+        result = cur->value;
+    }
+    return result;
+}
+
+int csh_putvar(const char *name, const char *value) {
+    struct csh_env_entry *existing = csh_get_env_entry(name);
+    if(NULL != existing) {
+        free(existing->value);
+        existing->value = strdup(value);
+    } else {
+        struct csh_env_entry *new_entry = calloc(1, sizeof(*new_entry));
+        new_entry->name = strdup(name);
+        new_entry->value = strdup(value);
+        SLIST_INSERT_HEAD(&csh_env, new_entry, list);
+    }
+    return 0;
+}
+
+int csh_delvar(const char *name) {
+    int res = 0;
+    struct csh_env_entry *var = csh_get_env_entry(name);
+    if(var) {
+        free(var->name);
+        free(var->value);
+        SLIST_REMOVE(&csh_env, var, csh_env_entry, list);
+        free(var);
+    } else {
+        res = 1;
+    }
+    return res;
+}
+
+void csh_clearenv() {
+    struct csh_env_entry *var;
+    while (!SLIST_EMPTY(&csh_env))
```

Comment:
Are we at the point where we want to enforce the placement of brackets?

Reply:
Oversight from my part, but bound to happen as long as we do not enforce automatic formatting, either at the IDE/editor level (clang-format or other) or as a git pre-commit hook. I'll fix that

Reply:
See commit 706666ee758854fec321e97fd9f57af788b63361